server/athenian/api/internal/miners/jira/issue.py

```diff
@@ -812,74 +816,150 @@ def _validate_and_clean_issues(df: pd.DataFrame, acc_id: int) -> pd.DataFrame:
-        prs: PullRequestFactsMap,
-        meta_ids: Tuple[int, ...],
-        mdb: DatabaseLike,
-    ) -> None:
-        """Load and insert "jira_id" to the PR facts."""
-        pr_node_ids = defaultdict(list)
-        for node_id, repo in prs:
-            pr_node_ids[node_id].append(repo)
-        jira_map = await cls.load_pr_jira_mapping(pr_node_ids, meta_ids, mdb)
-        for pr_node_id, jira in jira_map.items():
-            for repo in pr_node_ids[pr_node_id]:
-                try:
-                    prs[(pr_node_id, repo)].jira_ids = jira
-                except KeyError:
-                    # we removed this PR in JIRA filter
-                    continue
-
     @classmethod
     @sentry_span
-    async def load_pr_jira_mapping(
+    async def load(
         cls,
         prs: Collection[int],
-        meta_ids: Tuple[int, ...],
+        meta_ids: tuple[int, ...],
         mdb: DatabaseLike,
-    ) -> Dict[int, List[str]]:
-        """Fetch the mapping from PR node IDs to JIRA issue IDs."""
+    ) -> dict[int, PRJIRAMapping]:
+        """Load the JIRA mapping for each PR in `prs`."""
         nprji = NodePullRequestJiraIssues
+
         if len(prs) >= 100:
             node_id_cond = nprji.node_id.in_any_values(prs)
         else:
             node_id_cond = nprji.node_id.in_(prs)
-        rows = await mdb.fetch_all(
-            sql.select([nprji.node_id, Issue.key])
-            .select_from(
-                sql.outerjoin(
-                    nprji,
-                    Issue,
-                    sql.and_(
-                        nprji.jira_acc == Issue.acc_id,
-                        nprji.jira_id == Issue.id,
-                    ),
-                ),
-            )
-            .where(
-                sql.and_(
-                    node_id_cond,
-                    nprji.node_acc.in_(meta_ids),
-                ),
-            ),
+        cols = [nprji.node_id, Issue.key, Issue.priority_name, Issue.type, Issue.project_id]
+        from_clause = sa.outerjoin(
+            nprji, Issue, sa.and_(nprji.jira_acc == Issue.acc_id, nprji.jira_id == Issue.id),
+        )
+        stmt = (
+            sa.select(cols)
+            .select_from(from_clause)
+            .where(node_id_cond, nprji.node_acc.in_(meta_ids))
         )
-        result = defaultdict(list)
+
+        rows = await mdb.fetch_all(stmt)
+        return cls._load_from_db_rows(rows)
+
+    @classmethod
+    def load_from_df(self, df: pd.Dataframe) -> dict[int, PRJIRAMapping]:
+        """Load JIRA mappings from a Dataframe as found in PRDataFrames.jiras."""
+        mapping_builders: dict[int, _PRJIRAMappingBuilder] = defaultdict(_PRJIRAMappingBuilder)
+        for r in df.itertuples():
+            pr_id, issue_key = r[0]  # df index is (pr_id, issue_key)
+            mapping_builders[pr_id].import_values(issue_key, r.project_id, r.priority_name, r.type)
+        return {pr_id: builder.finalize() for pr_id, builder in mapping_builders.items()}
+
```

Comment:
OK, this is prohibitively slow and should be rewritten.
- itertuples is extremely slow
- Python object manipulation is slow
- this part of the code stand on the critical path and is performance critical

Uff, this is tough. All the code should be rewritten, but it's tricky to explain how.